### ISSUE TITLE
[TaxonomyBundle] Error when override TaxonTranslation

### ISSUE THREAD

**Issue by gperdomor** [Sylius/Sylius#3298]
Some time ago, i got this error <code> [Gedmo\Exception\InvalidMappingException] Unable to find tree parent slug relation through field - [parent] in class Sylius\Component\Taxonomy\Model\TaxonTranslation </code> when i tried to install <code>SyliusTaxonomy</code> version <code>v0.14.0</code>, then was solved with an update.

Now i get this error again, this time when i add a new field to <code>TaxonTranslation</code>, adding new fields to <code>TaxonomyTranslation</code> works fine, but not on <code>TaxonTranslation</code>. The schema update command is executed without errors, but not the <code>doctrine:fixtures:load</code> comand, also edit a taxon in the backend is impossible.

**Comment by michalmarcinkowski** [comment]
Maybe your issue is connected with https://github.com/Sylius/Sylius/issues/2294#issuecomment-72010609?

**Comment by MichaelKubovic** [comment]
I'm afraid I'm experiencing the same issue here. Although I'm using sylius standard edition, I've set sylius/sylius to dev-master in composer so I have the latest version.

```Fatal error: Uncaught exception 'Gedmo\Exception\InvalidMappingException' with message 'Unable to find tree parent slug relation through field - [parent] in class - Sylius\Component\Taxonomy\Model\TaxonTranslation' in /Users/michael/Sites/madweb/madad.sk/web/vendor/sylius/sylius/src/Sylius/Bundle/TranslationBundle/GedmoHandler/TranslationSlugHandler.php:139```

my config:
```yml
sylius_taxonomy:
    classes:
        taxon:
            model: Madness\MadadBundle\Entity\Taxon
            translation: 
                model: Madness\MadadBundle\Entity\TaxonTranslation
```

in my overriden Taxon.php
```php
public static function getTranslationClass()
{
    return 'Madness\MadadBundle\Entity\TaxonTranslation';
}
```

I've noticed that ```TranslationSlugHandler``` complains about mapping of ```Sylius\Component\Taxonomy\Model\TaxonTranslation``` which is weird as I've configured my override correctly.

Additionally, the fact that TranslationSlugHandler yields anything is weird as in my ```TaxonTranslation.orm.xml``` I have overridden permalink NOT to use this handler at all...

```xml
<entity name="Madness\MadadBundle\Entity\TaxonTranslation" table="sylius_taxon_translation">
        <attribute-overrides>
            <attribute-override name="permalink">
                <field name="permalink" column="permalink" type="string" unique="false">
                    <gedmo:slug fields="id, name" unique="false" />
                </field>
            </attribute-override>
        </attribute-overrides>
</entity>
```

Now I'm pretty sure I've configured everything correctly as there is little to screw there, so I'm kind of stuck with these symptoms... might be a bug?

**Comment by MichaelKubovic** [comment]
@michalmarcinkowski I believe this not a feature request but a real bug. I've managed to reproduce the problem on fresh sylius-standard clone with "sylius/sylius": "dev-master".

This is all you need to reproduce the issue:
https://github.com/MichaelKubovic/Sylius-Standard/commit/35e945bd2d9f7eb4fb28e888b200eef0bffd44a7

I'm more than happy to put together a PR if someone can lead me to the right direction....

**Comment by gperdomor** [comment]
I my case, i'm using Sylius 0.15.0

**Comment by pjedrzejewski** [comment]
Looks like a bug indeed, I have no clue so far where it can come from, but will investigate.

**Comment by khyor** [comment]
I had exactly the same problem till today !

I use sylius standalone  bundle with with a default configuration for taxonomy : 
```yml
sylius_taxonomy:
    classes:
        taxonomy:
            model: Sylius\Component\Taxonomy\Model\Taxonomy
            form:
                default: Sylius\Bundle\TaxonomyBundle\Form\Type\TaxonomyType
            translation:
                model: Sylius\Component\Taxonomy\Model\TaxonomyTranslation
                form:
                    default: Sylius\Bundle\TaxonomyBundle\Form\Type\TaxonomyTranslationType
        taxon:
            model: Sylius\Component\Taxonomy\Model\Taxon
            form:
                default: Sylius\Bundle\TaxonomyBundle\Form\Type\TaxonType
            translation:
                model:  Sylius\Component\Taxonomy\Model\TaxonTranslation
                form:
                    default: Sylius\Bundle\TaxonomyBundle\Form\Type\TaxonTranslationType
```

The only change made was in the `AppKernel` with the order of declaration bundle sylius (RessourceBundle before TranslationBundle) :
```php
    new Sylius\Bundle\TaxonomyBundle\SyliusTaxonomyBundle(),
    new Sylius\Bundle\ResourceBundle\SyliusResourceBundle(),
    new Sylius\Bundle\TranslationBundle\SyliusTranslationBundle(),
    new Sylius\Bundle\LocaleBundle\SyliusLocaleBundle(),
```

However, I had another error message with this order :  `[Doctrine\DBAL\Schema\SchemaException] There is no column with name 'locale' on table 'sylius_taxon_translation'.` 

For try, I commented out lines 41 to 43 in  `TaxonTranslation.orm.xml` where  `locale` was defined :
```xml
    <unique-constraints>
        <unique-constraint columns="locale,permalink" name="permalink_uidx" />
    </unique-constraints>
```

and the tables have been created in base after composer update.

However I notice that fields  `locale` and  `translatable_id`  are not present in the  `sylius_taxon_translation` and  `sylius_taxonomy_translation` tables like abstractTranslation wasn't mapped. an idea ?

I don't know if it's related or if it could help you find where does the problem !

**Comment by pjedrzejewski** [comment]
@khyor Hey Danny! Did you include DoctrineBundle after the rest of the Sylius bundles?

**Comment by khyor** [comment]
@pjedrzejewski Yes i did. I was careful on this !

**Comment by khyor** [comment]
I tested with my custom bundle `Igor/Bundle/BusinessBundle` with class `Job` and `JobTranslation`.
They simply extends your abstract Methods for Translation (without tree setup).
And Everything is working : Fields `locale` and `translatable_id` are present in my table.

**Comment by khyor** [comment]
@pjedrzejewski @gperdomor On the initial problem, I don't know if makes a real diffrence but in my project I added Gedmo_tree on the doctrine ORM configuration.

```yml
doctrine:
    orm:
        auto_generate_proxy_classes: %kernel.debug%
        entity_managers:
            default:
                auto_mapping: true
                mappings:
                    gedmo_loggable:
                        type: annotation
                        prefix: Gedmo\Loggable\Entity
                        dir: "%kernel.root_dir%/../vendor/gedmo/doctrine-extensions/lib/Gedmo/Loggable/Entity"
                        is_bundle: false
                    gedmo_tree:
                        type: annotation
                        prefix: Gedmo\Tree\Entity
                        dir: "%kernel.root_dir%/../vendor/gedmo/doctrine-extensions/lib/Gedmo/Tree/Entity"
                        is_bundle: false
```

**Comment by gperdomor** [comment]
@pjedrzejewski i see something weird, in <code>TranslationSlugHandler</code>, the
<code>validate</code> method calls !$meta->isSingleValuedAssociation(<code>$options['relationField']</code>) but the throw exception uses <code>$options['relationParentRelationField']</code>...

Anyway, change the parameter in the validation to use <code>$options['relationParentRelationField']</code> not solve the problem, but maybe is a good start point.

**Comment by steffenbrem** [comment]
I had the same issue and I resolved it by putting `SyliusResourceBundle` at the end of all other Sylius bundles in `AppKernel`.

**Comment by steffenbrem** [comment]
Yeah this definitely is a bug. I could run `php app/console doctrine:update:schema --force` now, but when  saving new taxons, it gives the same old error again. Everything seems to work fine when you comment the line @gperdomor described.

**Comment by steffenbrem** [comment]
The only way to currently extend the `TaxonTranslation` class is to copy its content from `Sylius\Bundle\TranslationsBundle` and the mapping too and use it for your own `TaxonTranslation`. Somehow the `TranslationSlugHandler` does not take mapping inheritance into account.

**Comment by mojzis** [comment]
i think there is also a slight issue in 
`Sylius/Component/Core/Model/Taxonomy.php`, the constructor is setting the root Taxon from the same namespace. 
```
$this->setRoot(new Taxon());
```
So anyone willing to override Taxon has to override Taxonomy as well :).

**Comment by michalmarcinkowski** [comment]
@mojzis this will be resolved after introducing Factories for object creation, look forward to it very soon :)

**Comment by khyor** [comment]
@MichaelKubovic  a date for the new version with Factories ? :)

**Comment by michalmarcinkowski** [comment]
@khyor factories are already merged (https://github.com/Sylius/Sylius/pull/3613), we just need to refactor the existing classes to remove the `new` in constructor.

**Comment by gperdomor** [comment]
@michalmarcinkowski mmmm... i don't know if i'm doing something wrong, but i'm still get the  <code>Unable to find tree parent slug relation through field - [parent] in class - Sylius\Component\Taxonomy\Model\TaxonTranslation</code> error

**Comment by pamil** [comment]
@gperdomor take a look at #3659, I had the same error today, it should work on the newest `master`.

**Comment by gperdomor** [comment]
@pamil thank you. I need wait the update in packagist, but in the meanwile, i will change the file manually for test.

**Comment by pamil** [comment]
I guess it can be closed (evolved into #3662 :)).
